In [1]:
#load libraries
import pandas as pd
import numpy as np
import seaborn as sns

In [2]:
#importing data
forest=pd.read_csv('D:\ExcelR\Assignments\Downloaded\SVM\\forestfires.csv')
forest.head()

month  day  FFMC   DMC     DC  ISI  temp  RH  wind  rain  ...  monthfeb  \
0   mar  fri  86.2  26.2   94.3  5.1   8.2  51   6.7   0.0  ...         0   
1   oct  tue  90.6  35.4  669.1  6.7  18.0  33   0.9   0.0  ...         0   
2   oct  sat  90.6  43.7  686.9  6.7  14.6  33   1.3   0.0  ...         0   
3   mar  fri  91.7  33.3   77.5  9.0   8.3  97   4.0   0.2  ...         0   
4   mar  sun  89.3  51.3  102.2  9.6  11.4  99   1.8   0.0  ...         0   

   monthjan  monthjul  monthjun  monthmar  monthmay  monthnov  monthoct  \
0         0         0         0         1         0         0         0   
1         0         0         0         0         0         0         1   
2         0         0         0         0         0         0         1   
3         0         0         0         1         0         0         0   
4         0         0         0         1         0         0         0   

   monthsep  size_category  
0         0          small  
1         0          small  
2         0          small  
3         0          small  
4         0          small  

[5 rows x 31 columns]

In [3]:
forest.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 517 entries, 0 to 516
Data columns (total 31 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   month          517 non-null    object 
 1   day            517 non-null    object 
 2   FFMC           517 non-null    float64
 3   DMC            517 non-null    float64
 4   DC             517 non-null    float64
 5   ISI            517 non-null    float64
 6   temp           517 non-null    float64
 7   RH             517 non-null    int64  
 8   wind           517 non-null    float64
 9   rain           517 non-null    float64
 10  area           517 non-null    float64
 11  dayfri         517 non-null    int64  
 12  daymon         517 non-null    int64  
 13  daysat         517 non-null    int64  
 14  daysun         517 non-null    int64  
 15  daythu         517 non-null    int64  
 16  daytue         517 non-null    int64  
 17  daywed         517 non-null    int64  
 18  monthapr  

-No null values, shape of data has 516 rows and 31 columns

In [4]:
fr=pd.DataFrame(forest)

In [5]:
FR=fr.iloc[:,6:11]

In [6]:
FR.head()

temp  RH  wind  rain  area
0   8.2  51   6.7   0.0   0.0
1  18.0  33   0.9   0.0   0.0
2  14.6  33   1.3   0.0   0.0
3   8.3  97   4.0   0.2   0.0
4  11.4  99   1.8   0.0   0.0

In [7]:
row_indexes=FR[FR['area']>=5].index

FR.loc[row_indexes,'Area']="large"

row_indexes=FR[FR['area']<5].index

FR.loc[row_indexes,'Area']="small"

FR.head()

temp  RH  wind  rain  area   Area
0   8.2  51   6.7   0.0   0.0  small
1  18.0  33   0.9   0.0   0.0  small
2  14.6  33   1.3   0.0   0.0  small
3   8.3  97   4.0   0.2   0.0  small
4  11.4  99   1.8   0.0   0.0  small

In [8]:
FR=FR.drop('area', axis=1)

In [9]:
FR.head()

temp  RH  wind  rain   Area
0   8.2  51   6.7   0.0  small
1  18.0  33   0.9   0.0  small
2  14.6  33   1.3   0.0  small
3   8.3  97   4.0   0.2  small
4  11.4  99   1.8   0.0  small

In [10]:
from sklearn.model_selection import train_test_split
import sklearn.svm as svm

In [11]:
X= FR.iloc[:,0:4]
y= FR.iloc[:,4]
X_train, X_test, y_train, y_test= train_test_split(X,y, test_size=0.3)

In [12]:
X_train

temp  RH  wind  rain
78   14.7  66   2.7   0.0
342  20.6  37   1.8   0.0
274   5.1  61   8.0   0.0
134  14.9  38   2.7   0.0
215  17.0  27   4.9   0.0
..    ...  ..   ...   ...
186  11.5  39   5.8   0.0
23   19.5  43   5.8   0.0
120  10.3  74   2.2   0.0
74   27.6  30   1.3   0.0
434  22.2  45   3.6   0.0

[361 rows x 4 columns]

In [13]:
X_test

temp  RH  wind  rain
479  22.6  57   4.9   0.0
146  13.2  40   5.4   0.0
483  31.0  27   5.4   0.0
332  17.1  41   2.2   0.0
231  27.8  27   3.1   0.0
..    ...  ..   ...   ...
485  30.6  28   3.6   0.0
242  15.4  66   4.0   0.0
331  21.5  28   4.5   0.0
350  10.1  75   3.6   0.0
58   13.9  40   5.4   0.0

[156 rows x 4 columns]

In [14]:
y_train

78     small
342    small
274    large
134    small
215    large
       ...  
186    large
23     small
120    small
74     small
434    small
Name: Area, Length: 361, dtype: object

In [15]:
y_test

479    large
146    small
483    small
332    large
231    large
       ...  
485    small
242    large
331    large
350    small
58     small
Name: Area, Length: 156, dtype: object

## Grid search CV

In [16]:
from sklearn.svm import SVC

In [17]:
# kernel = rbf
model_rbf = SVC(kernel = "rbf")
model_rbf.fit(X_train,y_train)
pred_test_rbf = model_rbf.predict(X_test)

np.mean(pred_test_rbf==y_test)

0.6602564102564102

In [18]:
from sklearn.metrics import classification_report, confusion_matrix
print(confusion_matrix(y_test,pred_test_rbf ))

[[  0  53]
 [  0 103]]


In [19]:
print(classification_report(y_test,pred_test_rbf))

              precision    recall  f1-score   support

       large       0.00      0.00      0.00        53
       small       0.66      1.00      0.80       103

    accuracy                           0.66       156
   macro avg       0.33      0.50      0.40       156
weighted avg       0.44      0.66      0.53       156



D:\Anaconda\lib\site-packages\sklearn\metrics\_classification.py:1245: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
D:\Anaconda\lib\site-packages\sklearn\metrics\_classification.py:1245: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
D:\Anaconda\lib\site-packages\sklearn\metrics\_classification.py:1245: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


## -As we can see there is 69% accuracy for kernal rbf

In [20]:
#Now let us see for other kernal as well


In [21]:
# kernel = linear
model_rbf = SVC(kernel = "linear")
model_rbf.fit(X_train,y_train)
pred_test_linear = model_rbf.predict(X_test)

np.mean(pred_test_linear==y_test)

0.6602564102564102

In [22]:
from sklearn.metrics import classification_report, confusion_matrix
print(confusion_matrix(y_test,pred_test_linear ))

[[  0  53]
 [  0 103]]


In [23]:
print(classification_report(y_test,pred_test_linear))

              precision    recall  f1-score   support

       large       0.00      0.00      0.00        53
       small       0.66      1.00      0.80       103

    accuracy                           0.66       156
   macro avg       0.33      0.50      0.40       156
weighted avg       0.44      0.66      0.53       156



D:\Anaconda\lib\site-packages\sklearn\metrics\_classification.py:1245: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
D:\Anaconda\lib\site-packages\sklearn\metrics\_classification.py:1245: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
D:\Anaconda\lib\site-packages\sklearn\metrics\_classification.py:1245: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


## Here the accuracy for kernal linear is 69% which is similar to kernal=rbf